src/3rdparty/walkontable/src/table.js

```diff
@@ -773,83 +698,87 @@ class Table {
     return this.TBODY.childNodes[this.rowFilter.sourceToRendered(row)];
   }
 
-  getFirstRenderedRow() {
-    return this.wot.wtViewport.rowsRenderCalculator.startRow;
-  }
-
-  getFirstVisibleRow() {
-    return this.wot.wtViewport.rowsVisibleCalculator.startRow;
-  }
-
-  getFirstRenderedColumn() {
-    return this.wot.wtViewport.columnsRenderCalculator.startColumn;
-  }
-
   /**
-   * @returns {Number} Returns -1 if no row is visible
+   * 0-based index of column header
+   * @param {Number} level
    */
-  getFirstVisibleColumn() {
-    return this.wot.wtViewport.columnsVisibleCalculator.startColumn;
+  isColumnHeaderLevelRendered(level) {
+    const columnHeaders = this.wot.getSetting('columnHeaders');
+    const columnHeadersCount = columnHeaders.length;
+    return level > (columnHeadersCount - 1);
   }
 
   /**
-   * @returns {Number} Returns -1 if no row is visible, otherwise source index of the last rendered row
+   * 0-based index of row header
+   * @param {Number} level
    */
-  getLastRenderedRow() {
-    return this.wot.wtViewport.rowsRenderCalculator.endRow;
+  isRowHeaderLevelRendered(level) {
+    const columnHeaders = this.wot.getSetting('rowHeaders');
+    const columnHeadersCount = columnHeaders.length;
+    return level > (columnHeadersCount - 1);
   }
 
-  /**
-   * @returns {Number} Returns source index of last visible row
-   */
-  getLastVisibleRow() {
-    return this.wot.wtViewport.rowsVisibleCalculator.endRow;
-  }
+  isRowBeforeRenderedRows(row) {
+    const first = this.getFirstRenderedRow();
+    if (row < 0) {
+      if (first === 0) {
```

Comment:
You can use `return first !== 0;` instead.

Reply:
But then the inline comment that is on the next line will be more difficult to comprehend. Do you think that I remove the comment?

Reply:
It shouldn't be `return first <= 0;`? 🤔

I think the comment would be useful when we back here next time.

Reply:
Good point. Actually, I think the logic in this function was good but messy. The comment was incorrect, though.

Anyway, I have simplified the logic and explained it in JSDoc: 0d527006fd5e79f2f135e218289722ab03edfdee